File: 0238-product-of-array-except-self/solution.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n=len(nums)
        result= [1] * n
        p=1
        for i in range(n):
            result[i] = p
            p *= nums[i]
        s=1
        for i in range(n-1,-1,-1):
            result[i] *= s
            s *= nums[i]
        return result              

File: 0238-product-of-array-except-self/solution.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int zero_count = 0;
        int prod = 1;
        for(int i : nums){
            if(i == 0){
                zero_count++;
                continue;
            }
            prod *= i;
        }
        for(int i = 0; i < nums.length; i++){
            if(zero_count > 1) {
                nums[i] = 0;
            }
            else if(zero_count == 1){
                if(nums[i] == 0) nums[i] = prod;
                else nums[i] = 0;
            }
            else{
                nums[i] = prod/ nums[i];
            }
        }
        return nums;
    }
}
